src/System.Private.CoreLib/src/System/Runtime/Intrinsics/X86/Enums.cs

```diff
@@ -167,23 +168,56 @@ public enum FloatComparisonMode : byte
         TrueUnorderedSignaling = 31,
     }
 
+    [Flags]
     public enum StringComparisonMode : byte {
+        /// <summary>
+        /// _SIDD_CMP_EQUAL_ANY
+        /// </summary>
         EqualAny = 0x00,
+
+        /// <summary>
+        /// _SIDD_CMP_RANGES
+        /// </summary>
         Ranges = 0x04,
+
+        /// <summary>
+        /// _SIDD_CMP_EQUAL_EACH
+        /// </summary>
         EqualEach = 0x08,
+
+        /// <summary>
+        /// _SIDD_CMP_EQUAL_ORDERED
+        /// </summary>
         EqualOrdered = 0x0c,
-        NegativePolarity = 0x10,
-        MaskedNegativePolarity = 0x30,
+
+        /// <summary>
+        /// _SIDD_NEGATIVE_POLARITY
+        /// </summary>
+        NegativeResult = 0x10,
+
+        /// <summary>
+        /// _SIDD_MASKED_NEGATIVE_POLARITY
+        /// </summary>
+        NegativeUsefulResult = 0x30,
+
+        /// <summary>
+        /// _SIDD_LEAST_SIGNIFICANT
+        /// </summary>
         LeastSignificant = 0x00,
+
+        /// <summary>
+        /// _SIDD_MOST_SIGNIFICANT
+        /// </summary>
         MostSignificant = 0x40,
-    }
 
+        /// <summary>
+        /// _SIDD_BIT_MASK
+        /// </summary>
+        BitMask = 0x00,
 
-    public enum ResultsFlag : byte {
-        CFlag = 0,
-        NotCFlagAndNotZFlag = 1,
-        OFlag = 2,
-        SFlag = 3,
-        ZFlag = 4,
+        /// <summary>
+        /// _SIDD_UNIT_MASK
+        /// </summary>
+        UnitMask = 0x40,
```

Comment:
How are `UnitMask` and `MostSignificant` (or `LeastSignificant` and `BitMask`) distinguished? I don't think we should have multiple names for the same value in a single enum - are these used in different contexts? If so, I would think they'd be different enums.

Reply:
> are these used in different contexts?

Yes.
`UnitMask` and `BitMask` is only used for `CompareMask` (`PCMPESTRM`).
`LeastSignificant` and `MostSignificant` is only used for `CompareIndex` (`PCMPESTRI`).

Reply:
> If so, I would think they'd be different enums.

So, `CompareMask` and `CompareIndex` would have an additional imm parameter.
How about encoding these two flags into the intrinsic functiom name?

Reply:
> How about encoding these two flags into the intrinsic function name?

I don't think that's necessary - I just don't see why the enums should be the same.

Reply:
I think it depends on whether we want to map closely with the underlying hardware instruction, which I thought was a desire.

Section `4.1.5 Output Selection` in https://www.intel.com/content/dam/www/public/us/en/documents/manuals/64-ia-32-architectures-software-developer-vol-2b-manual.pdf

![image](https://user-images.githubusercontent.com/8291187/41362551-451ca21a-6ef7-11e8-9001-23e5c0f77d72.png)

Since it will be a `Flags` enum (after this PR), I don't think this is a huge issue. It follows the underlying HW instruction design, and is understandable - use Least|Most Significant with the `CompareIndex` method, use Bit|Unit Mask with the `CompareMask` method.

Reply:
@eerhardt, I think @CarolEidt's point is that it is still consistent with the underlying instructions. But is exposing it to the user in a more "safe" and "understandable" manner.

You can certainly do `Bit|Unit`  with CompareIndex (instead of `Least|Most`) and it will do the "correct" thing, but it would make the code harder to understand/etc.

Reply:
> You can certainly do Bit|Unit with CompareIndex (instead of Least|Most) and it will do the "correct" thing, but it would make the code harder to understand/etc.

Yes, but there is a trade-off between safer/stronger enum type-checking and usability. I think it may not worthwhile to complicate the APIs just for these two flags.

Reply:
@fiigii, is it really that more complex?

On the API side, we'll have two enums, that are restricted to the Index vs Mask overloads (which we already have).

On the JIT side, we can just ensure that we carry it around as a byte constant and don't need to worry about there being two enums.

Reply:
> On the API side, we'll have two enums,
> On the JIT side, we can just ensure that we carry it around as a byte constant

Don't you mean define `LeastSignificant`/`BitMask` as two enum types? 

```csharp
    public enum StringComparisonMode : byte {
        // ...
    }

    public enum IndexMode : byte {
        // ...
        /// <summary>
        /// _SIDD_LEAST_SIGNIFICANT
        /// </summary>
        LeastSignificant = 0x00,

        /// <summary>
        /// _SIDD_MOST_SIGNIFICANT
        /// </summary>
        MostSignificant = 0x40,
    }

    public enum MaskMode : byte {
        // ...
        /// <summary>
        /// _SIDD_BIT_MASK
        /// </summary>
        BitMask = 0x00,

        /// <summary>
        /// _SIDD_UNIT_MASK
        /// </summary>
        UnitMask = 0x40,
    }
```

If so, these APIs would need an additional imm-parameter, right?

Reply:
> If so, these APIs would need an additional imm-parameter, right?

Right.

We might also be able to just double the number of exposed methods and encode the "mode" in the name, if @eerhardt and @CarolEidt thought that was more reasonable.

Reply:
>> I think @CarolEidt's point is that it is still consistent with the underlying instructions

>If so, these APIs would need an additional imm-parameter

> We might also be able to just double the number of exposed methods and encode the "mode" in the name

At what point does it become "inconsistent" with the underlying instructions?

Reply:
> At what point does it become "inconsistent" with the underlying instructions?

IMO, it would be if we no longer guaranteed the particular instruction would be emitted or when we no longer made it possible to cause certain functionality to happen (like if we excluded one of the possible flag values).

Reply:
I think that exposing something like `GetIndexOfFirstMatch` is completely within the realm of matching the underlying instruction and it mirrors other intrinsic APIs where we are doing similar things (like `Sse.Compare*OrderedScalar`, where all 6+ overloads are exposed, rather than providing an enum that the user passes in for the Comparisonmode).

Reply:
Finally, these two solutions complicate API design (@eerhardt 's solution) or JIT implementation (@tannergooding's solution, especially for non-const fallback). So, I said that may not worthwhile for such bit type safety.

Reply:
How about we do the following?  @tannergooding or @CarolEidt, if you want the enum to change, can you open a new corefx issue with an API proposal for changing the enum? That will probably be the best way forward. Designing it in this PR probably isn't the best solution.

Reply:
> Using an enum allows me to use the API without making the choice.

I'm not sure how valid this is. The choice can make a fairly big difference on how you write your algorithm, so you will generally need to make the choice and be aware of how it impacts your code, right out.

(You will also have to specify one of the two enum values either way)

Reply:
> I'm not sure how valid this is.

It's completely valid when I'm exploring the API to learn how it works.

> (You will also have to specify one of the two enum values either way)

Only in the 3 enums design.  As currently proposed I wouldn't need to specify `Bit/UnitMask` when calling `CompareMask`.

Reply:
> As currently proposed I wouldn't need to specify Bit/UnitMask when calling CompareMask.

I'm not convinced that's a good thing 😄

Reply:
That is, these APIs and instructions are already fairly hard to understand. Making parts of it more obvious, even if it results in us exposing double the methods might be a good thing (It looks like we have ~50 methods exposed right now).